### PR TITLE
Foundation for User Uploads

### PR DESCRIPTION
**Pull request by Navbryce**
## What is this?

Uses Microsoft Azure for User Uploads
### Developer relevant information
**Handling user upload**:
1. Pass the file to the `uploadFile` function
2. Save the `blobPath` returned by the function inside the database

**Displaying uploaded file**:
1. With the `blobPath`, call `linkToUploadedFile(blobPath)` to generate a URL to the image. This can be used in an `Image` tag for example

See `pages/user-uploads-example.tsx` for an example of uploading and displaying uploaded images. You must be logged in for uploading to work
### Implementation details
**Uploading a file request path  (Checks auth for uploading and then sends URL that the user can upload to):**
1. User requests to upload a file
2. We generate a URL that lets a user upload a file from their browser to the following locaiton `{USER_ID}/{FILE_NAME}`. This URL is valid for 10 minutes, and can only be used for uploads 
3. The user uses this URL to upload the file from their browser

**Fetching an uploaded file request path (Checks auth for file accesses):**
1. Checks to see if the first part of the path matches the user id of the individual logged in. If not, the file access is invalid
2. The request is piped to microsoft azure. This should put relatively minimal load on our server

We can make any changes as necessary to how we authorize uploads and accesses.

## Setup

- `npm install`
- Two new environment variables: `AZURE_STORAGE_ACCOUNT` and `AZURE_STORAGE_KEY`

### CODE REVIEW

File: server/utils/azure-storage.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+export const blobService = new BlobServiceClient(
+  `https://${
+    process.env.AZURE_STORAGE_ACCOUNT as string
+  }.blob.core.windows.net`,
+  new StorageSharedKeyCredential(
+    process.env.AZURE_STORAGE_ACCOUNT as string,
+    process.env.AZURE_STORAGE_KEY as string
+  )
+);
+export const containerSerivce = blobService.getContainerClient(
```
Comment (mjbernet3) [inline server/utils/azure-storage.ts:19]: Spelling on this export.  I know, I am OCD.

File: src/actions/User.ts [hunk 2]
```diff
@@ -56,3 +60,55 @@ export function chapterUpdateUser(userUpdate: ChapterUpdateUser) {
...
+ * @param nameWithoutExtension - the name to give the file in storage (without the extension). if two files have the same name
+ * for a user, the file will overwrite the existing file. if a name is not
+ * provided, a randomly generated name will be used that does NOT collide with an
+ * existing file
+ * @param onProgress - called on progress events. on fast networks, this
+ * will might just emit one event
+ *
+ * Will throw error if file is too large
+ */
+export async function uploadFile(
```
Comment (mjbernet3) [inline src/actions/User.ts:76]: I think it makes more sense to move this function to your uploaded-files.ts util file.  Let me know if you disagree though.

File: server/utils/azure-storage.ts (hunk 1 above)
Comment (Navbryce) [reply server/utils/azure-storage.ts:19]: Good catch!

File: src/actions/User.ts (hunk 2 above)
Comment (Navbryce) [reply src/actions/User.ts:76]: Yep, that makes sense to me

**Review by mjbernet3 — APPROVED** [review]

**Pull request merged**